[FEAT] 태그레벨 데이터 뷰에서 DB 데이터로 보여주기

giantim opened the pull request:
- 지금 태그 레벨을 수치화해서 보여줄 때 모습이 다음과 같습니다.
![image](https://user-images.githubusercontent.com/39546083/90957040-61947c00-e4c6-11ea-84e6-a2d698c0cfbf.png)

그림을 보면은 각 태그 레벨 별 최소 해시태그 수 만 필요로 하고 있고 마지막 레벨에 대해서는 아무 값도 필요하지 않습니다.
왜냐하면 html로 저 막대를 만들었을 때 가장 앞과 가장 뒤에는 글씨를 넣을 수 없기 때문입니다.
지금 TagLevelResponse를 만들었을 때는 태그 레벨, 최소 해시태그 수, 최대 해시태그 수를 모두 담아서 응답을 보내도록 설계를 했습니다.
여러분들은 사용하는 값만 응답에 포함시켜서 보내는 것이 더 좋다고 생각하는지 아니면 지금 구조도 상관 없다고 생각하는지 의견을 듣고싶습니다 :)

- 버그를 하나 발견했는데 이거는 설명을 잘 못하겠어서 월요일에 만나서 보여드리도록 하겠습니다 😂

closed: #178

giantim committed:
feat: tag-level api 작성

giantim committed:
feat: tagLevelQuery 서비스 구현 및 테스트

giantim committed:
feat: 뷰에서 태그 레벨 정보를 데이터베이스에 저장된 값으로 보여지게 변경

- 해시태그 수 계산 로직 공통 부분 추출
- 프론트 파일 스타일 수정
- 태그 레벨 응답값에 레벨 값 추가 및 문서화 작성

DunDung reviewed (approved):
고생하셨습니다 라빈 👍 
포맷팅 된 코드가 많네용

ordinCode commented on hashtagmap-web/src/test/java/com/songpapeople/hashtagmap/service/TagLevelQueryServiceTest.java:
```diff
@@ -0,0 +1,47 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class TagLevelQueryServiceTest {
+    @Mock
+    private TagLevelRepository tagLevelRepository;
+
+    private TagLevelQueryService tagLevelQueryService;
+
+    @BeforeEach
+    private void setUp() {
+        tagLevelQueryService = new TagLevelQueryService(tagLevelRepository);
+    }
+
+    @DisplayName("태그 레벨을 5개 가져오는지 테스트")
+    @Test
+    void findTagLevelsTest() {
+        List<TagLevel> tagLevels = Arrays.asList(
+                new TagLevel(1L, 100L, 105L),
+                new TagLevel(2L, 106L, 110L),
+                new TagLevel(3L, 111L, 115L),
+                new TagLevel(4L, 116L, 120L),
+                new TagLevel(5L, 121L, 125L)
+        );
+        when(tagLevelRepository.findAll()).thenReturn(tagLevels);
+
+        List<TagLevelResponse> tagLevelResponses = tagLevelQueryService.findTagLevels();
+
+        assertThat(tagLevelResponses).hasSize(5);
+    }
```
지금은 repository에서 실제로 데이터를 가져오는걸 테스트하는게 없는거 같은데, service에서는 repository를 mock치지 않고 실제로 찔러보는 테스트를 해야할 것 같습니다.

giantim replied:
단순히 repository에서 findAll 메서드를 테스트를 해야 할까요?
저는 서비스에서는 응답값을 만들어서 주는 역할을 테스트 하는것으로 충분하다고 생각이 듭니다 :)

pci2676 replied:
저는 이 테스트가 그렇게 유의미해 보이지 않아요. mock으로 5개를 뱉도록 만들고 5개를 가져오는지 테스트 한다라는게 큰 의미를 갖지 못하는 것 같네요!
차라리 tagLevelRepository에서 findAll() 메서드를 사용하는게 아니라 findTop5ByOrderByCreatedDate() 같은 메서드를 사용하는게 더 의미있을것 같아요 위 테스트 코드만 보고는 findAll()이 5개만 가져올거라는 보장을 해줄수 없는 것 같거든요!
그래서 findTop5ByOrderByCreatedDate()과 같이 limit 5 가 걸린 쿼리문을 날리는 형식의 통합테스트가 필요할 것 같습니다.

giantim replied:
findAll 메서드가 아니라 querydsl을 이용해서 비즈니스 로직에 맞는 메서드를 만드는게 더 적합해 보이네요!

ordinCode commented on hashtagmap-web/src/main/java/com/songpapeople/hashtagmap/dto/TagLevelResponse.java:
```diff
@@ -0,0 +1,32 @@
+                .mapToObj(index -> {
+                    TagLevel tagLevel = tagLevels.get(index);
+                    return new TagLevelResponse(tagLevel.getId(),
+                            tagLevel.getMinHashtagCount(), tagLevel.getMaxHashtagCount());
```
메서드로 분리하면 어때요?

pci2676 replied:
저는 그냥 for문을 사용하는게 더 보기 편할것 같다는 생각도 드네요

ordinCode reviewed (changes requested):
라빈 수고했어요~!
코멘트남겼으니 확인해주세요ㅎㅎ

moonyoungCHAE commented on hashtagmap-core/src/main/java/com/songpapeople/hashtagmap/taglevel/model/TagLevel.java:
```diff
@@ -39,9 +39,9 @@ public void update(Long minHashtagCount, Long maxHashtagCount) {
+            String detailMessage = String.format("최소 해시태그 개수(%s)는 최대 해시태그 개수(%s)보다 많을 수 없습니다.",
                     minHashtagCount, maxHashtagCount);
-            throw new CoreException(CoreExceptionStatus.INVALID_TAG_LEVEL, detailMessgae);
+            throw new CoreException(CoreExceptionStatus.INVALID_TAG_LEVEL, detailMessage);
```
오타 있었군요 ㅋㅋ 바꿔줘서 고맙습니다~

moonyoungCHAE commented on hashtagmap-web/front/src/components/detail-modal/DetailModal.vue:
```diff
@@ -62,7 +63,8 @@ export default {
     makeHashtagCount() {
-      return (this.getDetailModal.hashtagCount / 1000).toFixed(1);
+      return convertHashtagCount(this.getDetailModal.hashtagCount);
+      // return (this.getDetailModal.hashtagCount / 1000).toFixed(1);
```
```suggestion
```

giantim replied:
홀리~ 고마워요 :)

moonyoungCHAE commented on hashtagmap-web/front/src/components/tag-level/TagLevelDetail.vue:
```diff
@@ -24,22 +32,31 @@
   }),
 
   methods: {
+    ...mapActions(["setTagLevelDetails"]),
```
이 부분 없어도 되지 않나요?!

giantim replied:
```
async created() {
    await this.setTagLevelDetails();
  },
```
여기서 호출해야 해서 필요합니다 :)

moonyoungCHAE reviewed (approved):
라빈~ 수고하셨습니다.
👍
응답은 지금 구조로 가도 될 것 같아요. 변경될 일이 없을 것 같진 않지만 그럴 경우를 고려해서 최소,최대 다 받아오는 게 좋은 것 같아요.

pci2676 commented on hashtagmap-web/front/vue.config.js:
```diff
@@ -5,18 +5,18 @@ module.exports = {
     proxy: {
-      '/': {
+      "/": {
         target: "http://localhost:9000",
```
이왕 포맷팅 하는 김에 9000도 proxyPort 변수로 바꿔주시는건 어떨까요?

pci2676 reviewed (changes requested):
학성이 리뷰한 부분에 대해서 제 생각을 남겼습니다~
확인해주세요~

giantim committed:
refactor: 리뷰 반영

- TagLevel을 가져올 때 가장 최신의 5개만 가져오도록 변경
- TagLevelResponse 로직 수정
- DetailModal의 주석 제거
